<a href="https://colab.research.google.com/github/Krish5986/Project_Samarth_Q-A/blob/main/EDA_crop_production.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [ ]:
from google.colab import drive
drive.mount("/content/drive")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
agri_df = pd.read_csv("/content/drive/MyDrive/My_Projects/Build_For_Bharat/crop_production_data.csv")
agri_df.head()

state_name district_name  crop_year  season               crop   area_  \
0  Uttar Pradesh        MEERUT       1997  Kharif          Arhar/Tur   986.0   
1  Uttar Pradesh        MEERUT       1997  Kharif              Bajra    27.0   
2  Uttar Pradesh        MEERUT       1997  Kharif              Jowar    50.0   
3  Uttar Pradesh        MEERUT       1997  Kharif              Maize  4556.0   
4  Uttar Pradesh        MEERUT       1997  Kharif  Moong(Green Gram)    14.0   

   production_  
0        693.0  
1         38.0  
2          4.0  
3       6817.0  
4          6.0

In [ ]:
agri_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36640 entries, 0 to 36639
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   state_name     36640 non-null  object 
 1   district_name  36640 non-null  object 
 2   crop_year      36640 non-null  int64  
 3   season         36640 non-null  object 
 4   crop           36640 non-null  object 
 5   area_          36640 non-null  float64
 6   production_    35905 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 2.0+ MB


In [ ]:
agri_df.isnull().sum()

state_name         0
district_name      0
crop_year          0
season             0
crop               0
area_              0
production_      735
dtype: int64

##Findings
We have 735 NaN values

In [ ]:
zeros_production = (agri_df['production_'] == 0).sum()
zeros_production

np.int64(610)

## Finding

we have 610 values of production which is equivalent to 0

# Rename the columns names

In [ ]:
column_lst = agri_df.columns.tolist()
column_lst

['state_name',
 'district_name',
 'crop_year',
 'season',
 'crop',
 'area_',
 'production_']

##Findings
We have two columns whose name is not good as per the data analysis and does not seems to be good

area_, production_

In [ ]:
agri_df.rename(columns = {'area_':'area','production_':'production'},inplace = True)
print("Columns renamed successfully")

Columns renamed successfully


In [ ]:
#Recheck the columns name in the dataset
agri_df.head()

state_name district_name  crop_year  season               crop    area  \
0  Uttar Pradesh        MEERUT       1997  Kharif          Arhar/Tur   986.0   
1  Uttar Pradesh        MEERUT       1997  Kharif              Bajra    27.0   
2  Uttar Pradesh        MEERUT       1997  Kharif              Jowar    50.0   
3  Uttar Pradesh        MEERUT       1997  Kharif              Maize  4556.0   
4  Uttar Pradesh        MEERUT       1997  Kharif  Moong(Green Gram)    14.0   

   production  
0       693.0  
1        38.0  
2         4.0  
3      6817.0  
4         6.0

## Handle missing production values


In [ ]:
#drop rows NaN values
agri_df_cleaned = agri_df.dropna(subset = ['production']).copy()

In [ ]:
#Recheck about the null values
print(f"New shape: {agri_df_cleaned.shape}")

New shape: (35905, 7)


In [ ]:
agri_df_cleaned.isnull().sum()

state_name       0
district_name    0
crop_year        0
season           0
crop             0
area             0
production       0
dtype: int64

#Unique Values

In [ ]:
print("Unique State Names:")
print(agri_df_cleaned['state_name'].unique())

Unique State Names:
['Uttar Pradesh' 'Madhya Pradesh' 'Maharashtra' 'Karnataka' 'Gujarat'
 'Punjab' 'West Bengal' 'Rajasthan' 'Haryana' 'Andhra Pradesh']


In [ ]:
#Check unique district names (sample only first 20 to keep it readable)
print("\nUnique District Names (sample):")
print(agri_df_cleaned['district_name'].unique()[:20])



Unique District Names (sample):
['MEERUT' 'AGRA' 'LUCKNOW' 'GHAZIABAD' 'KANPUR NAGAR' 'VARANASI' 'ALIGARH'
 'MORADABAD' 'GORAKHPUR' 'ALLAHABAD' 'HOSHANGABAD' 'VIDISHA' 'SEHORE'
 'BHOPAL' 'INDORE' 'JABALPUR' 'GWALIOR' 'UJJAIN' 'SAGAR' 'RATLAM']


In [ ]:
# Check seasons
print("\nUnique Seasons:")
print(agri_df_cleaned['season'].unique())


Unique Seasons:
['Kharif' 'Rabi' 'Summer' 'Whole Year' 'Autumn' 'Winter']


In [ ]:
#Check crops
print("\nUnique Crops (sample):")
print(agri_df_cleaned['crop'].unique())


Unique Crops (sample):
['Arhar/Tur' 'Bajra' 'Jowar' 'Maize' 'Moong(Green Gram)' 'Oilseeds total'
 'Onion' 'Rice' 'Sesamum' 'Small millets' 'Total foodgrain' 'Urad'
 'Barley' 'Gram' 'Masoor' 'Peas & beans (Pulses)' 'Rapeseed &Mustard'
 'Sunflower' 'Wheat' 'Potato' 'Sugarcane' 'Cotton(lint)' 'Castor seed'
 'Turmeric' 'Groundnut' 'Sannhamp' 'Dry chillies' 'Garlic' 'Sweet potato'
 'Coriander' 'Other Kharif pulses' 'Other  Rabi pulses' 'Guar seed'
 'Linseed' 'Moth' 'Soyabean' 'Jute' 'Dry ginger' 'Ragi' 'Tobacco' 'Banana'
 'Paddy' 'Safflower' 'Niger seed' 'Beans & Mutter(Vegetable)' 'Bhindi'
 'Brinjal' 'Cabbage' 'Cauliflower' 'Citrus Fruit' 'Mango'
 'Other Fresh Fruits' 'Other Vegetables' 'Papaya' 'Pome Fruit' 'Tomato'
 'other misc. pulses' 'Orange' 'Other Citrus Fruit' 'Water Melon'
 'Horse-gram' 'Khesari' 'Cowpea(Lobia)' 'Mesta' 'Grapes' 'Pulses total'
 'Other Cereals & Millets' 'other oilseeds' 'Black pepper' 'Arecanut'
 'Cardamom' 'Coconut' 'Arcanut (Processed)' 'Atcanut (Raw)'
 'Cashew

#Standardize District Names

In [ ]:
#district_name to Title case

agri_df_cleaned['district_name'] = agri_df_cleaned['district_name'].str.title()


In [ ]:
#Check
print(agri_df_cleaned['district_name'].unique()[:5])

['Meerut' 'Agra' 'Lucknow' 'Ghaziabad' 'Kanpur Nagar']


In [ ]:
#Clean 'Crop' Names

#Strip leading and trailing whitespace
agri_df_cleaned['corp'] = agri_df_cleaned['crop'].str.strip()

In [ ]:
#Convert to Title Case for consitency
agri_df_cleaned['crop'] = agri_df_cleaned['crop'].str.title()

In [ ]:
print("Cleaned Crop Names (sample): ")
print(agri_df_cleaned['crop'].unique()[:5])

Cleaned Crop Names (sample): 
['Arhar/Tur' 'Bajra' 'Jowar' 'Maize' 'Moong(Green Gram)']


In [ ]:
#Standardize 'crop' names using a mapping dictionary

# Key = Current non-standard value, Value = Desired standard value
crop_mapping = {
    'Arhar/Tur': 'Arhar',                   # Resolve slash
    'Moong(Green Gram)': 'Moong',           # Remove parentheses
    'Urad': 'Black Gram',                   # Use common name
    'Masoor': 'Lentil',                     # Use common name
    'Peas & Beans (Pulses)': 'Peas & Beans',# Remove extra context
    'Paddy': 'Rice',                        # Standardize 'Paddy' (unprocessed) to 'Rice'
    'Dry Chillies': 'Chillies',
    'Dry Ginger': 'Ginger',
    'Sweet Potato': 'Sweetpotato',           # Optional: standardize compound word spacing
    'Other Kharif Pulses': 'Other Pulses',
    'Other Rabi Pulses': 'Other Pulses',
    'Other Misc. Pulses': 'Other Pulses',
    'Pulses Total': 'Pulses Total',         # Keep totals distinct
    'Cashewnut Processed': 'Cashewnut',     # Standardize Cashewnut entries
    'Cashewnut Raw': 'Cashewnut',
    'Atcanut (Raw)': 'Arecanut',            # Fix typo/variant
    'Other Oilseeds': 'Oilseeds Total',     # Grouping
    'Small Millets': 'Millets Total',
    'Other Cereals & Millets': 'Millets Total',
    'Horse-Gram': 'Horse Gram',
    'Cowpea(Lobia)': 'Cowpea',
    'Other Fresh Fruits': 'Other Fruits',
    'Pome Granet': 'Pomegranate',           # Fix spelling
    'Peas  (Vegetable)': 'Peas',             # Clean up extra spaces and context
    'Beans & Mutter(Vegetable)': 'Beans & Mutter',
    'Cotton(Lint)' : 'Cotton',
    'Rapeseed &Mustard': 'Rapeseed & Mustard',
    'Other  Rabi Pulses': 'Other Pulses',
    'Arcanut (Processed)': 'Arecanut'



}

# The .map() function replaces values based on the dictionary.
agri_df_cleaned['crop'] = agri_df_cleaned['crop'].replace(crop_mapping)


# Verification Check
print("\nStandardized Crop Names (sample after mapping):")
print(agri_df_cleaned['crop'].unique())


Standardized Crop Names (sample after mapping):
['Arhar' 'Bajra' 'Jowar' 'Maize' 'Moong' 'Oilseeds Total' 'Onion' 'Rice'
 'Sesamum' 'Millets Total' 'Total Foodgrain' 'Black Gram' 'Barley' 'Gram'
 'Lentil' 'Peas & Beans' 'Rapeseed & Mustard' 'Sunflower' 'Wheat' 'Potato'
 'Sugarcane' 'Cotton' 'Castor Seed' 'Turmeric' 'Groundnut' 'Sannhamp'
 'Chillies' 'Garlic' 'Sweetpotato' 'Coriander' 'Other Pulses' 'Guar Seed'
 'Linseed' 'Moth' 'Soyabean' 'Jute' 'Ginger' 'Ragi' 'Tobacco' 'Banana'
 'Safflower' 'Niger Seed' 'Beans & Mutter' 'Bhindi' 'Brinjal' 'Cabbage'
 'Cauliflower' 'Citrus Fruit' 'Mango' 'Other Fruits' 'Other Vegetables'
 'Papaya' 'Pome Fruit' 'Tomato' 'Orange' 'Other Citrus Fruit'
 'Water Melon' 'Horse Gram' 'Khesari' 'Cowpea' 'Mesta' 'Grapes'
 'Pulses Total' 'Black Pepper' 'Arecanut' 'Cardamom' 'Coconut' 'Cashewnut'
 'Tapioca' 'Other Fibres' 'Korra' 'Samai' 'Bottle Gourd' 'Cucumber'
 'Lemon' 'Sapota' 'Pomegranate' 'Peas' 'Varagu']


In [ ]:
#Final whitespace strip on 'state_name' and 'season'

# Strip whitespace from 'state_name'
agri_df_cleaned['state_name'] = agri_df_cleaned['state_name'].str.strip()

# Strip whitespace from 'season'
agri_df_cleaned['season'] = agri_df_cleaned['season'].str.strip()


# Final Verification
print("\nFinal Check for Cleaned Unique Seasons:")
print(agri_df_cleaned['season'].unique())

print("\nFinal Check for Cleaned Unique States:")
print(agri_df_cleaned['state_name'].unique())


Final Check for Cleaned Unique Seasons:
['Kharif' 'Rabi' 'Summer' 'Whole Year' 'Autumn' 'Winter']

Final Check for Cleaned Unique States:
['Uttar Pradesh' 'Madhya Pradesh' 'Maharashtra' 'Karnataka' 'Gujarat'
 'Punjab' 'West Bengal' 'Rajasthan' 'Haryana' 'Andhra Pradesh']


In [ ]:
agri_df_cleaned['season'].value_counts()

season
Kharif        15294
Rabi          10558
Whole Year     7229
Summer         2514
Autumn          182
Winter          128
Name: count, dtype: int64

In [ ]:
agri_df_cleaned.to_csv('/content/drive/MyDrive/My_Projects/Build_For_Bharat/agri_df.csv',index = False)